pkg/service/backup_handler.go

```diff
@@ -55,118 +56,111 @@ func NewBackupHandler(config *model.Config, backupPolicy *model.BackupPolicy) (*
 
-// scheduleFullBackup runs the full backup periodically.
-func (h *BackupHandler) scheduleFullBackup(ctx context.Context) {
-	ticker := time.NewTicker(time.Duration(*h.backupPolicy.IntervalMillis) * time.Millisecond)
-	defer ticker.Stop()
-loop:
-	for {
-		select {
-		case now := <-ticker.C:
-			if isStaleTick(now) {
-				slog.Error("Skipped full backup", "name", *h.backupPolicy.Name)
-				backupSkippedCounter.Inc()
-				break
-			}
-			// read the state first and check
-			state := h.backend.readState()
-			if !h.isFullEligible(now, state.LastRun) {
-				slog.Debug("The full backup is not due to run yet", "name", *h.backupPolicy.Name)
-				break
-			}
-			if !h.fullBackupInProgress.CompareAndSwap(false, true) {
-				slog.Debug("Backup is currently in progress, skipping full backup", "name", *h.backupPolicy.Name)
-				break
-			}
-			backupRunFunc := func() {
-				backupService.BackupRun(h.backupPolicy, h.cluster, h.storage, shared.BackupOptions{})
-			}
-			out := stdIO.Capture(backupRunFunc)
-			util.LogCaptured(out)
-			slog.Debug("Completed full backup", "name", *h.backupPolicy.Name)
-
-			// increment backupCounter metric
-			backupCounter.Inc()
-
-			// update the state
-			h.updateBackupState(now, state)
-
-			// clean incremental backups
-			h.backend.CleanDir(model.IncrementalBackupDirectory)
-
-			// release the lock
-			h.fullBackupInProgress.Store(false)
+// Schedule schedules backup for the defining policy.
+func (h *BackupHandler) Schedule(ctx context.Context) {
+	slog.Info("Scheduling full backup", "name", *h.backupPolicy.Name)
+	h.scheduleBackupPeriodically(ctx, h.runFullBackup)
 
-		case <-ctx.Done():
-			slog.Debug("ctx.Done in scheduleFullBackup")
-			break loop
-		}
+	if h.backupPolicy.IncrIntervalMillis != nil && *h.backupPolicy.IncrIntervalMillis > 0 {
+		slog.Info("Scheduling incremental backup", "name", *h.backupPolicy.Name)
+		h.scheduleBackupPeriodically(ctx, h.runIncrementalBackup)
 	}
-	slog.Info("Exiting scheduling loop for full backup", "name", *h.backupPolicy.Name)
 }
 
-// scheduleBackup runs the incremental backup periodically.
-func (h *BackupHandler) scheduleIncrementalBackup(ctx context.Context) {
-	ticker := time.NewTicker(time.Duration(*h.backupPolicy.IncrIntervalMillis) * time.Millisecond)
-	defer ticker.Stop()
-loop:
-	for {
-		select {
-		case now := <-ticker.C:
-			if isStaleTick(now) {
-				slog.Error("Skipped incremental backup", "name", *h.backupPolicy.Name)
-				incrBackupSkippedCounter.Inc()
-				break
-			}
-			// read the state first and check
-			state := h.backend.readState()
-			if state.LastRun == (time.Time{}) {
-				slog.Debug("Skip incremental backup until initial full backup is done", "name", *h.backupPolicy.Name)
-				break
+// scheduleBackupPeriodically runs the backup periodically based on the provided interval.
+func (h *BackupHandler) scheduleBackupPeriodically(
+	ctx context.Context,
+	backupFunc func(time.Time)) {
+	go func() {
+		ticker := time.NewTicker(1000 * time.Millisecond)
```

Comment:
Make the tick duration a constant (potentially will be a configuration property).